From 7ea25aab559bdc158173e1ab47f4746c52fa3b5f Mon Sep 17 00:00:00 2001
From: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
Date: Mon, 29 Jun 2026 08:02:43 +0000
Subject: [PATCH] fs: sysfs: Fix reference leak in
 sysfs_break_active_protection()

jira VULN-980
cve CVE-2024-26993
commit-author Alan Stern <stern@rowland.harvard.edu>
commit a90bca2228c0646fc29a72689d308e5fe03e6d78

The sysfs_break_active_protection() routine has an obvious reference
leak in its error path.  If the call to kernfs_find_and_get() fails then
kn will be NULL, so the companion sysfs_unbreak_active_protection()
routine won't get called (and would only cause an access violation by
trying to dereference kn->parent if it was called).  As a result, the
reference to kobj acquired at the start of the function will never be
released.

Fix the leak by adding an explicit kobject_put() call when kn is NULL.

	Signed-off-by: Alan Stern <stern@rowland.harvard.edu>
Fixes: 2afc9166f79b ("scsi: sysfs: Introduce sysfs_{un,}break_active_protection()")
	Cc: Bart Van Assche <bvanassche@acm.org>
	Cc: stable@vger.kernel.org
	Reviewed-by: Bart Van Assche <bvanassche@acm.org>
	Acked-by: Tejun Heo <tj@kernel.org>
Link: https://lore.kernel.org/r/8a4d3f0f-c5e3-4b70-a188-0ca433f9e6f9@rowland.harvard.edu
	Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
(cherry picked from commit a90bca2228c0646fc29a72689d308e5fe03e6d78)
	Signed-off-by: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
---
 fs/sysfs/file.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/fs/sysfs/file.c b/fs/sysfs/file.c
index 9aefa7779b29..2a52f9df0b27 100644
--- a/fs/sysfs/file.c
+++ b/fs/sysfs/file.c
@@ -440,6 +440,8 @@ struct kernfs_node *sysfs_break_active_protection(struct kobject *kobj,
 	kn = kernfs_find_and_get(kobj->sd, attr->name);
 	if (kn)
 		kernfs_break_active_protection(kn);
+	else
+		kobject_put(kobj);
 	return kn;
 }
 EXPORT_SYMBOL_GPL(sysfs_break_active_protection);
-- 
2.52.0

